doc: user guide on contributing to NCS

martiles opened the pull request:
Added a user guide how developers can contribute to NCS.

Signed-off-by: Martin Lesund <martin.lesund@nordicsemi.no>

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+and test your changes thoroughly before submitting.
+
+The general GitHub workflow used by NCS developers uses a combination of
+command line Git commands and browser interaction with GitHub.  As it is with
```
As usual with Git...

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+.. _Create a Fork of NCS manifest repo (fw-nrfconnect-nrf):
+  https://github.com/NordicPlayground/fw-nrfconnect-nrf#fork-destination-box
+
+#. `Create a Fork of fw-nrfconnect-nrf`_
```
Wrong text for the link above?

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+
+     git remote rename origin upstream
+
+   Let Git know about the fork you just created, naming it ``origin``::
```
Do this before renaming the remote above, makes sense to create it before "using" it

martiles replied:
I am a bit unsure on how you want this setup in the doc.

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+#. Create a topic branch (off of master) for your work (if you're addressing
+   an issue, we suggest including the issue number in the branch name)::
+
+     git checkout master
```
remove this line

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+   an issue, we suggest including the issue number in the branch name)::
+
+     git checkout master
+     git checkout -b fix_comment_typo
```
`git checkout -b fix_comment_typo upstream/master`

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+#. Push your topic branch with your changes to your fork in your personal
+   GitHub account::
+
+     git push origin fix_comment_typo
```
This should be:
`git push -u origin fix_comment_typo`
so that the branch is tracked after

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+
+     git push origin fix_comment_typo
+
+#. In your web browser, go to your forked repo and click on the
```
Alternatively, use the URL that GitHub sends you back after the `git push` above

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+   can create another branch to work on another issue. (Be sure to make your
+   new branch off of master and not the previous branch.)::
+
+     git checkout master
```
remove this line

carlescufi commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,127 @@
+   new branch off of master and not the previous branch.)::
+
+     git checkout master
+     git checkout -b fix_another_issue
```
`git checkout -b fix_another_issue upstream/master`

martiles commented:
Thanks for the comments @carlescufi.
Please check the new commit

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:92: ERROR:TRAILING_WHITESPACE: trailing whitespace
#92: FILE: doc/nrf/ug_contribution.rst:86:
+#. In your web browser, go to your forked repo or use the URL that Github sends $

-:117: ERROR:TRAILING_WHITESPACE: trailing whitespace
#117: FILE: doc/nrf/ug_contribution.rst:111:
+^I $

-:119: ERROR:TRAILING_WHITESPACE: trailing whitespace
#119: FILE: doc/nrf/ug_contribution.rst:113:
+^I$

-:121: ERROR:TRAILING_WHITESPACE: trailing whitespace
#121: FILE: doc/nrf/ug_contribution.rst:115:
+^I$

- total: 4 errors, 0 warnings, 126 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

ru-fu commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,142 @@
+Contributing to the |NCS|
+*************************
+
+The |NCS| is maintained by Nordic Semiconductor, but it is open to community contributions.
```
Is this okay as introduction?

ru-fu commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,142 @@
+As usual with Git, there are multiple ways of getting a task done.
+The following steps show a typical workflow:
+
+1. :ref:`dm-wf-fork`.
```
I added a link to the documentation we already have for this ( https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_dev_model.html#fork-a-repository-of-the-ncs ). Not sure if we want to keep the substeps - they are a bit simpler than the full instructions.
I changed "upstream" to "ncs" though - I think it's a really bad idea to tell people to call their NCS upstream "upstream" in one set of instructions and "ncs" in another ...

martiles replied:
I agree, good idea 👍

ru-fu commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,142 @@
+
+      git checkout -b *branch_name* ncs/master
+
+   If you are addressing a specific issue, we suggest including the issue number in the branch name.
```
We don't have public issues in NCS (yet), or do we?

martiles replied:
That is correct, I will remove that sentence.

ru-fu commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,142 @@
+   If you are addressing a specific issue, we suggest including the issue number in the branch name.
+
+#. Make changes, test locally, change, test, test again, ...
+   Check out the *sanitycheck* section in Zephyr's :ref:`zephyr:Contribution Tools` documentation as well.
```
Does it make sense to mention sanitycheck? And if so, is this the correct link? (https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/contribute/index.html?highlight=sanitycheck#sanitycheck)

ru-fu commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,142 @@
+     git commit -s
+
+   The ``-s`` option automatically adds your ``Signed-off-by:`` to your commit message.
+   This signature indicates that you agree with the :ref:`zephyr:DCO`.
```
Do we use the same DCO as Zephyr? Does it make sense to link?

ru-fu commented on doc/nrf/ug_contribution.rst:
```diff
@@ -0,0 +1,142 @@
+   The ``-s`` option automatically adds your ``Signed-off-by:`` to your commit message.
+   This signature indicates that you agree with the :ref:`zephyr:DCO`.
+   Without this line, your commit will be rejected.
+   See the *Commit Guidelines* section in Zephyr's :ref:`zephyr:Contribution workflow` for specific guidelines for writing your commit messages.
```
We don't have commit guidelines for NCS - does it make sense to link to Zephyr's?

mbolivar commented on doc/nrf/ug_contribution.rst:
```diff
@@ -1,126 +1,142 @@
+   .. parsed-literal::
+      :class: highlight
+
+      git add *files*
```
This patch has removed the note about using `git add -p` and advocates a single `git add`.

This is in direct conflict with the guidance `As a general practice, we encourage to make small, controlled changes`. Here, "small, controlled changes" means a series of commits which each "do one thing" instead of a single git add + git commit which mixes together diff from several logically distinct changes to the source code.

The latter style of development is harder to review, and harder to bisect.

I think this documentation should continue to describe (or at least hint as it's doing now) that you can use things like `git add -p` or other tools which can stage individual hunks to split the diff into logical commits.

Here is the relevant part of the Linux documentation about contributions which spells this out in more detail:

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#separate-your-changes

ru-fu replied:
Remember that this guide is aimed at beginners, so we should describe the easiest way to do something, and not give alternatives (because people won't know which is the best alternative for them to use).
I've never used ``add -p`` - is it easier than just listing the files (if you put yourself in the shoes of someone who is NOT familiar with git)? I tried it out now and it ignored untracked files.

mbolivar replied:
> Remember that this guide is aimed at beginners, so we should describe the easiest way to do something, and not give alternatives (because people won't know which is the best alternative for them to use).

In my personal opinion, if you are going to contribute to NCS, knowledge of how to structure your git commits for ease of review is an important skill, and we should expect that level of professionalism from potential contributors.

This isn't about separating changes from separate files into separate commits, it's about setting up your commits so they each do one logical thing. Using `git add -p` is useful if your working tree contains changes in one file that should logically be part of two distinct commits, because it allows you to stage individual hunks from each file you're adding to the index. There are other ways to accomplish the goal -- as this patch itself said, " As usual with Git, there are multiple ways of getting a task done" -- but this is a common recommendation since it's built into git.

@carlescufi do you have any thoughts on this?

martiles commented:
Squashed the commit by @ru-fu (thanks 👍 ), and addressed some other small changes.
@carlescufi could you please go through the remaining questions if we should link to zephyrs guidelines, sanity check etc. ?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=1873) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=1873) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nrfconnect/sdk-nrf?pullRequest=1873) it.</sub>

carlescufi commented:
Note: This is an automated message.
@martiles this Pull Request has not been updated for more than a month, please close it if not applicable anymore.
